File: backend/src/main/java/com/ecomate/backend/entity/RecyclingCentre.java
package com.ecomate.backend.entity;

import jakarta.persistence.*;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "recycling_centres")
public class RecyclingCentre {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "officer_id")
    private User officer;

    @Column(name = "officer_email", nullable = false)
    private String officerEmail;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    private String address;

    @Column(nullable = false)
    private String city;

    @Column(name = "latitude")
    private Double latitude = 6.9271;

    @Column(name = "longitude")
    private Double longitude = 79.8612;

    @Column(name = "distance_km")
    private Double distanceKm = 1.2;

    @Column(name = "contact_number", nullable = false)
    private String contactNumber;

    @Column(nullable = false)
    private String email;

    @Column(name = "operating_hours", nullable = false)
    private String operatingHours = "Mon - Sat: 8:00 AM - 5:30 PM";

    @Column(name = "is_open", nullable = false)
    private Boolean isOpen = true;

    @Column(length = 1000)
    private String notes = "";

    @OneToMany(mappedBy = "recyclingCentre", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<RecyclingCentreMaterial> centreMaterials = new ArrayList<>();

    public RecyclingCentre() {
    }

    public RecyclingCentre(User officer, String officerEmail, String name, String address, String city,
                           String contactNumber, String email, String operatingHours, Boolean isOpen, String notes) {
        this.officer = officer;
        this.officerEmail = officerEmail;
        this.name = name;
        this.address = address;
        this.city = city;
        this.contactNumber = contactNumber;
        this.email = email;
        this.operatingHours = operatingHours;
        this.isOpen = isOpen;
        this.notes = notes;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public User getOfficer() {
        return officer;
    }

    public void setOfficer(User officer) {
        this.officer = officer;
    }

    public String getOfficerEmail() {
        return officerEmail;
    }

    public void setOfficerEmail(String officerEmail) {
        this.officerEmail = officerEmail;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getAddress() {
        return address;
    }

    public void setAddress(String address) {
        this.address = address;
    }

    public String getCity() {
        return city;
    }

    public void setCity(String city) {
        this.city = city;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }

    public Double getDistanceKm() {
        return distanceKm;
    }

    public void setDistanceKm(Double distanceKm) {
        this.distanceKm = distanceKm;
    }

    public String getContactNumber() {
        return contactNumber;
    }

    public void setContactNumber(String contactNumber) {
        this.contactNumber = contactNumber;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getOperatingHours() {
        return operatingHours;
    }

    public void setOperatingHours(String operatingHours) {
        this.operatingHours = operatingHours;
    }

    public Boolean getIsOpen() {
        return isOpen;
    }

    public void setIsOpen(Boolean isOpen) {
        this.isOpen = isOpen;
    }

    public String getNotes() {
        return notes;
    }

    public void setNotes(String notes) {
        this.notes = notes;
    }

    public List<RecyclingCentreMaterial> getCentreMaterials() {
        return centreMaterials;
    }

    public void setCentreMaterials(List<RecyclingCentreMaterial> centreMaterials) {
        this.centreMaterials = centreMaterials;
    }
}

File: mobile/lib/models/recycling_centre.dart
class RecyclingCentre {
  final String id;
  final String? officerEmail;
  final int? officerId;
  final String name;
  final String address;
  final String city;
  final double distanceKm;
  final String contactNumber;
  final String email;
  final String operatingHours;
  final bool isOpen;
  final List<String> acceptedMaterials;
  final List<String> unsupportedMaterials;
  final String notes;

  const RecyclingCentre({
    required this.id,
    this.officerEmail,
    this.officerId,
    required this.name,
    required this.address,
    required this.city,
    required this.distanceKm,
    required this.contactNumber,
    required this.email,
    required this.operatingHours,
    required this.isOpen,
    required this.acceptedMaterials,
    required this.unsupportedMaterials,
    required this.notes,
  });

  RecyclingCentre copyWith({
    String? id,
    String? officerEmail,
    int? officerId,
    String? name,
    String? address,
    String? city,
    double? distanceKm,
    String? contactNumber,
    String? email,
    String? operatingHours,
    bool? isOpen,
    List<String>? acceptedMaterials,
    List<String>? unsupportedMaterials,
    String? notes,
  }) {
    return RecyclingCentre(
      id: id ?? this.id,
      officerEmail: officerEmail ?? this.officerEmail,
      officerId: officerId ?? this.officerId,
      name: name ?? this.name,
      address: address ?? this.address,
      city: city ?? this.city,
      distanceKm: distanceKm ?? this.distanceKm,
      contactNumber: contactNumber ?? this.contactNumber,
      email: email ?? this.email,
      operatingHours: operatingHours ?? this.operatingHours,
      isOpen: isOpen ?? this.isOpen,
      acceptedMaterials: acceptedMaterials ?? this.acceptedMaterials,
      unsupportedMaterials: unsupportedMaterials ?? this.unsupportedMaterials,
      notes: notes ?? this.notes,
    );
  }

  factory RecyclingCentre.fromJson(Map<String, dynamic> json) {
    return RecyclingCentre(
      id: json['id']?.toString() ?? '',
      officerEmail: json['officerEmail'] as String?,
      officerId: json['officerId'] is int ? json['officerId'] as int : null,
      name: json['name'] as String? ?? '',
      address: json['address'] as String? ?? '',
      city: json['city'] as String? ?? '',
      distanceKm: (json['distanceKm'] is num) ? (json['distanceKm'] as num).toDouble() : 1.2,
      contactNumber: json['contactNumber'] as String? ?? '',
      email: json['email'] as String? ?? '',
      operatingHours: json['operatingHours'] as String? ?? 'Mon - Sat: 8:00 AM - 5:30 PM',
      isOpen: json['isOpen'] as bool? ?? true,
      acceptedMaterials: (json['acceptedMaterials'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      unsupportedMaterials: (json['unsupportedMaterials'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      notes: json['notes'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'address': address,
      'city': city,
      'contactNumber': contactNumber,
      'email': email,
      'operatingHours': operatingHours,
      'isOpen': isOpen,
      'acceptedMaterials': acceptedMaterials,
      'unsupportedMaterials': unsupportedMaterials,
      'notes': notes,
    };
  }
}

